registration/include/pcl/registration/impl/transformation_estimation_symmetric_point_to_plane_lls.hpp

```diff
@@ -0,0 +1,251 @@
+pcl::registration::TransformationEstimationSymmetricPointToPlaneLLS<PointSource, PointTarget, Scalar>::
+estimateRigidTransformation (ConstCloudIterator<PointSource>& source_it, ConstCloudIterator<PointTarget>& target_it, Matrix4 &transformation_matrix) const
+{
+  using Vector6d = Eigen::Matrix<double, 6, 1>;
+  using Matrix6d = Eigen::Matrix<double, 6, 6>;
+
+  Matrix6d ATA;
+  Vector6d ATb;
+  ATA.setZero ();
+  ATb.setZero ();
+
+  // Approximate as a linear least squares problem
+  while (source_it.isValid () && target_it.isValid ())
+  {
+    if (!std::isfinite (source_it->x) ||
+        !std::isfinite (source_it->y) ||
+        !std::isfinite (source_it->z) ||
+        !std::isfinite (target_it->x) ||
+        !std::isfinite (target_it->y) ||
+        !std::isfinite (target_it->z) ||
+        !std::isfinite (source_it->normal_x) ||
+        !std::isfinite (source_it->normal_y) ||
+        !std::isfinite (source_it->normal_z) ||
+        !std::isfinite (target_it->normal_x) ||
+        !std::isfinite (target_it->normal_y) ||
+        !std::isfinite (target_it->normal_z))
+    {
+      ++target_it;
+      ++source_it;    
+      continue;
+    }
+
+    const float & sx = source_it->x;
+    const float & sy = source_it->y;
+    const float & sz = source_it->z;
+    const float & dx = target_it->x;
+    const float & dy = target_it->y;
+    const float & dz = target_it->z;
+    const float & nx = source_it->normal[0] + target_it->normal[0];
+    const float & ny = source_it->normal[1] + target_it->normal[1];
+    const float & nz = source_it->normal[2] + target_it->normal[2];
+
+    double a = nz*sy - ny*sz;
+    double b = nx*sz - nz*sx; 
+    double c = ny*sx - nx*sy;
+   
+    //    0  1  2  3  4  5
+    //    6  7  8  9 10 11
+    //   12 13 14 15 16 17
+    //   18 19 20 21 22 23
+    //   24 25 26 27 28 29
+    //   30 31 32 33 34 35
+   
+    ATA.coeffRef (0) += a * a;
+    ATA.coeffRef (1) += a * b;
+    ATA.coeffRef (2) += a * c;
+    ATA.coeffRef (3) += a * nx;
+    ATA.coeffRef (4) += a * ny;
+    ATA.coeffRef (5) += a * nz;
+    ATA.coeffRef (7) += b * b;
+    ATA.coeffRef (8) += b * c;
+    ATA.coeffRef (9) += b * nx;
+    ATA.coeffRef (10) += b * ny;
+    ATA.coeffRef (11) += b * nz;
+    ATA.coeffRef (14) += c * c;
+    ATA.coeffRef (15) += c * nx;
+    ATA.coeffRef (16) += c * ny;
+    ATA.coeffRef (17) += c * nz;
+    ATA.coeffRef (21) += nx * nx;
+    ATA.coeffRef (22) += nx * ny;
+    ATA.coeffRef (23) += nx * nz;
+    ATA.coeffRef (28) += ny * ny;
+    ATA.coeffRef (29) += ny * nz;
+    ATA.coeffRef (35) += nz * nz;
+
+    double d = nx*dx + ny*dy + nz*dz - nx*sx - ny*sy - nz*sz;
+    ATb.coeffRef (0) += a * d;
+    ATb.coeffRef (1) += b * d;
+    ATb.coeffRef (2) += c * d;
+    ATb.coeffRef (3) += nx * d;
+    ATb.coeffRef (4) += ny * d;
+    ATb.coeffRef (5) += nz * d;
+
+    ++target_it;
+    ++source_it;    
+  }
+  ATA.coeffRef (6) = ATA.coeff (1);
```

Comment:
A loop to copy might be more terse (or a triangular view to copy, but as per benchmarks, loop is a little faster and comparable to rolled out)

Reply:
Switched to a selfadjointView  and Cholesky solve to avoid the copy to the lower triangular half altogether.